### ISSUE TITLE
Beringei for IoT related sensors data

### ISSUE THREAD

**Issue by bu2** [facebookincubator/beringei#54]
Dear Beringei team,

I am exploring the usage of Beringei as a core data storage of time series in the context of Data Science / Machine Learning applied to IoT (sensors data).

Below, I introduce the story with my needs, constraints and questions. I would love to have any feedbacks on it !

- I need to handle hours of sensors data (up to 10 hours) at a time

- I have plenty of sensors, all with specific data rates (1, 2, 4, 8, 16... up to 2048Hz in the most common cases, but I can also have to handle high frequency signals: 44kHz, 66kHz)
    => so I would love to have nanosecond precision with adaptative time resolution per time serie (for example: I could chose seconds, millisecond or nanosecond precision, depending of the sensor, in hope it help the delta-compression...)

- today, a coherent dataset of sensors is about a few TB (a million of time series), but I could be interested to work on group of datasets and this could easily lead to multi PB scale

- the long term persistent storage is already handled and not part of the story

- I need the best possible performance for:
    - scanning time series (after resampling or not)
    - applying threshold detection
    - applying simple statistical computation (rolling mean/median/standard deviation...)
    - computing derived time series from any combination of other time series and store the result

- I am working in Python and weighting the opportunity to develop a Cython wrapper for Beringei

- I would love to read and write time series through Pandas dataframes

- I would love to be highly compatible with PySpark API (while being significantly more efficient than the Hadoop/Spark/Java stuff...). In my current understanding, Beringei does not cover this distributed computing aspect but you may have ideas to share

- I would love to be compatible with HPC technologies: MPI (distributed multi-processing), OpenMP (multi-thread), CUDA, SSE/AVX instruction sets. Those technologies would be used on the consumer side (reading/writing/processing jobs) but would require thread-safety built into Beringei and stringent memory alignment.

Few questions:
- Does Beringei could fit my story ?
- How does Gorilla/Beringei concept compare to BTrDB ?
- Is the auto-correlation feature described in Gorilla's paper available in Beringei ?
- if the answer at the first question is not no, is there anything in my story that could be of interest for Facebook (in term of priority) ?

Thank you for open sourcing Beringei !

Best Regards,

Bruno LE HYARIC.

**Comment by jteller** [comment]
Your use-case does not appear to be well supported (out of the box) by Beringei. For one discussion about the granularity of timestamps, you can refer to issue #21 .

A big thing about Beringei is scope of the project -- read @scottfranklin's comment in issue #14 for a point of view on how Beringei works. The main tl;dr is that it is amazing in terms of query latency and throughput, but that it has very few features (only simple put/get methods). It's targeted to be used as an add-on to an existing monitoring solution.

@at15 has posted a really cool comparison of different TSDBs here: https://github.com/xephonhq/awesome-time-series-database

**Comment by bu2** [comment]
Thank for the feedback, it is much appreciated.

I understand that my use case would require significant work but actually I am evaluating if Beringei could be a good starting point or if I need to go for a clean sheet design.

The things I like in Beringei (and Gorilla's paper) are:

- it's written in C/C++ and so it is Python-friendly

- the core data types in the current implementation seems to be already capable to handle high resolution timestamps and double precision floating point values ():
```
struct TimeValuePair {
  1: i64 unixTime,
  2: double value,
}
```

- I love the 12x data size reduction in memory and hope to substain that score on timeseries of different time resolution with a reasonable amount of code (I am no data compression specialist)

- I am especially interested by the narrow scope of Beringei. I intend to use it as a sort of cache for large dataset of timeseries and I hope to have the best performance for data exploration and analysis


You may advise on the following point:

- my datasets are already stored on cold storage, and I would need to load them in Beringei in bulk import. Then when exploring/processing the data, I would need to scan the time series efficiently through bulk reads. It seems I would have to implement those bulk I/O ?


Best Regards,

Bruno LE HYARIC.

**Comment by jteller** [comment]
Your list makes sense, but be aware:

Beringei, out of the box, cannot handle high resolution timestamps. unixTime is expressed in seconds, and it internally enforces both that it is storing data in realtime (it only allows data to be stored from now to some time period in the past), and that data points for a single time series are no closer together than 1x every 15 seconds. To support what you need will require some hacking.

Additionally, the 15x compression relies on some assumptions on how data your data is shaped (frequency of timestamps and "stability" of the values). It wouldn't be too difficult to change the assumptions to fit your data shape - but might also require some hacking.

Yes - you would need to implement those bulk I/O methods, because the main putData() method only allows points with a timestamp close to now (+/- a few minutes at the most). While a completely "new" code path, it would likely be easiest to add (see existing code for loading data from disk at startup: https://github.com/facebookincubator/beringei/blob/master/beringei/lib/DataBlockReader.cpp#L26-L107)

**Comment by bu2** [comment]
I see,

Actually I am planning to work on a quick & dirty prototype in the following weeks to evaluate the feasibility regarding my use case. I will keep you updated about the result.

Thanks again for your inputs !

**Comment by bu2** [comment]
Hi,

I have been diving into beringei source code for a while :)

As now, I have been able to write a time series with nanosecond precision and read some of its data back. But I faced some difficulties to unplug all the temporal logic specific to the original use case of Beringei : 
- timestamp behind/after,
- all the uint32_t that come In the "timestamp to bucket to timestamp roundtrip"

It was the opportunity to learn a lot a bot sharding and bucketing time series and to discover the folly::MPMCQueue.

Though I am joining your advice that beringei might be too far from my use case, I will try to infuse some of its design principles in my stack. I will let you known if I can come up with any tangible result.

Regards,

Bruno.
